go-controller/pkg/ovn/obj_retry.go

```diff
@@ -170,94 +171,134 @@ func (r *retryObjs) initRetryObjWithDelete(obj interface{}, key string, config i
-// then the function will try to create a new retryEntry in the map if , otherwise it will return nil.
-func (r *retryObjs) ensureRetryEntryLocked(key string, newRetryEntry *retryObjEntry) *retryObjEntry {
-	r.retryMutex.Lock()
-	entry := r.entries[key]
-	if entry != nil {
-		// do not hold on retryMutex while waiting on retryEntry to Lock
-		r.retryMutex.Unlock()
+func (r *retryObjs) deleteRetryEntry(key string) {
+	r.entriesMapMutex.Lock()
+	defer r.entriesMapMutex.Unlock()
+	delete(r.entries, key)
+}
 
-		// take the lock now
-		entry.Lock()
+func (r *retryObjs) loadRetryEntry(key string) (value *retryObjEntry, loaded bool) {
+	r.entriesMapMutex.Lock()
+	defer r.entriesMapMutex.Unlock()
+	value, ok := r.entries[key]
+	return value, ok
+}
 
-		// Check that the retryEntry wasn't deleted while we were waiting for its lock
-		r.retryMutex.Lock()
-		_, ok := r.entries[key]
-		if ok {
-			r.retryMutex.Unlock()
-			return entry
-		}
-		// since the entry went missing from the map, fallback below to assign the newRetryEntry
-		// with retryMutex locked from above
+func (r *retryObjs) loadOrStoreRetryEntry(key string, value *retryObjEntry) (actual *retryObjEntry, loaded bool) {
+	r.entriesMapMutex.Lock()
+	defer r.entriesMapMutex.Unlock()
+	if actual, ok := r.entries[key]; ok {
+		return actual, true
+	} else {
+		r.entries[key] = value
+		return value, false
+	}
+}
+
+func (r *retryObjs) loadRetryEntryWithLock(key string) (entry *retryObjEntry, loaded bool) {
+	entry, ok := r.loadRetryEntry(key)
+	if !ok {
+		return nil, false
+	}
+	entry.Lock()
+	// make sure entry wasn't deleted while we waited for lock
+	if _, loaded = r.loadRetryEntry(key); loaded {
+		return entry, true
+	} else {
 		entry.Unlock()
-		entry = nil
+		return nil, false
+	}
+}
+
+func (r *retryObjs) getRetryEntryKeys() []string {
+	r.entriesMapMutex.Lock()
+	defer r.entriesMapMutex.Unlock()
+	keys := make([]string, len(r.entries))
+	i := 0
+	for k := range r.entries {
+		keys[i] = k
+		i++
 	}
+	return keys
+}
 
-	if newRetryEntry != nil {
-		entry = newRetryEntry
-		r.entries[key] = entry
-		// this is fine since no one knows about this newRetryEntry yet
+// return a locked retryEntry, that was either loaded by existing key, or added with newRetryEntry value
+func (r *retryObjs) ensureRetryEntryLocked(key string, newRetryEntry *retryObjEntry) *retryObjEntry {
+	// if we add new entry, lock it before adding to prevent other threads getting entry Lock before this function returns
+	newRetryEntry.Lock()
+	entry, loaded := r.loadOrStoreRetryEntry(key, newRetryEntry)
+	if loaded {
+		// entry exists, try to Lock it
 		entry.Lock()
+		// check if entry was deleted while we waited for lock
+		// and add newRetryEntry if that's the case
+		updatedEntry, loaded := r.loadOrStoreRetryEntry(key, newRetryEntry)
+		if loaded {
+			// entry we loaded and locked is still in the map, return it
+			// unlock our tmp newRetryEntry, since we didn't use it
+			newRetryEntry.Unlock()
+			return entry
```

Comment:
That's a good observation - and all 3 files (egressgw, namespace and old ensureRetryEntryLocked) implement it differently. I like the egressgw option with .deleted field, but we need to create a single interface - because what we have now is not reusing the code, but copying it - which as we can see results in having 3 different implementations
egressgw uses .deleted https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/egressgw.go#L88
namespace compares objects https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/namespace.go#L449
and old ensureRetryEntryLocked ignores this (as current PR) https://github.com/ovn-org/ovn-kubernetes/blob/master/go-controller/pkg/ovn/obj_retry.go#L243